File: sdk/kotlin/src/main/kotlin/email/testinbox/client/internal/transport/Transport.kt
package email.testinbox.client.internal.transport

import email.testinbox.client.TestInboxApiException
import email.testinbox.client.TestInboxAuthException
import email.testinbox.client.TestInboxConflictException
import email.testinbox.client.TestInboxCredentialAlreadyCreatedException
import email.testinbox.client.TestInboxIdempotencyConflictException
import email.testinbox.client.TestInboxIdempotencyInProgressException
import email.testinbox.client.TestInboxForbiddenException
import email.testinbox.client.TestInboxInboxGoneException
import email.testinbox.client.TestInboxNotFoundException
import email.testinbox.client.TestInboxQuotaExceededException
import email.testinbox.client.TestInboxRateLimitException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import kotlinx.coroutines.future.await
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * INTERNAL transport layer (ADR-014): a thin, hand-written client over the
 * committed OpenAPI contract. Nothing in this package is part of the public
 * SDK surface and it may change without notice.
 */
@Serializable
internal data class CreateInboxRequestDto(
    val addressMode: String? = null,
    val ttlSeconds: Long? = null,
    val aliasHint: String? = null,
    val localPart: String? = null,
)

@Serializable
internal data class InboxDto(
    val id: String,
    val address: String,
    val addressMode: String = "GENERATED",
    val state: String = "ACTIVE",
    val createdAt: String? = null,
    val expiresAt: String? = null,
)

@Serializable
internal data class HeaderDto(val name: String, val value: String = "")

@Serializable
internal data class LinkDto(val href: String, val text: String? = null)

@Serializable
internal data class AttachmentDto(
    val id: String,
    val fileName: String? = null,
    val contentType: String? = null,
    val sizeBytes: Long = 0,
)

@Serializable
internal data class MessageDto(
    val id: String,
    val inboxId: String,
    val receivedAt: String? = null,
    val envelopeFrom: String? = null,
    val envelopeTo: String? = null,
    val parseStatus: String = "OK",
    val parseError: String? = null,
    val from: String? = null,
    val fromHeader: String? = null,
    val toHeader: String? = null,
    val subject: String? = null,
    val textBody: String? = null,
    val htmlBody: String? = null,
    val headers: List<HeaderDto> = emptyList(),
    val links: List<LinkDto> = emptyList(),
    val attachments: List<AttachmentDto> = emptyList(),
    val contentFingerprint: String? = null,
    val possibleDuplicateOfMessageId: String? = null,
    val rawSizeBytes: Long = 0,
)

@Serializable
internal data class MessagePageDto(val items: List<MessageDto> = emptyList(), val nextCursor: String? = null)

@Serializable
internal data class HeaderMatcherDto(val name: String, val value: String? = null)

@Serializable
internal data class MatcherDto(
    val from: String? = null,
    val subjectContains: String? = null,
    val subjectEquals: String? = null,
    val headers: List<HeaderMatcherDto> = emptyList(),
)

@Serializable
internal data class WaitRequestDto(val matcher: MatcherDto, val timeoutSeconds: Long)

@Serializable
internal data class WaitResultDto(
    val status: String = "TIMEOUT",
    val message: MessageDto? = null,
    val elapsedMs: Long = 0,
    val arrivedButUnmatchedCount: Int? = null,
    val parseFailedCount: Int? = null,
)

@Serializable
internal data class ApiKeyDto(
    val id: String,
    val publicId: String,
    val name: String? = null,
    val scopes: List<String> = emptyList(),
    val createdAt: String? = null,
    val expiresAt: String? = null,
    val revokedAt: String? = null,
    val lastUsedAt: String? = null,
    val createdByApiKeyId: String? = null,
)

/**
 * The wire shape of the one response that carries a credential. The secret
 * sits beside the metadata rather than inside it, which is what lets the
 * public [email.testinbox.client.ApiKeyMetadata] type have no field for it at
 * all (ADR-032 §4).
 */
@Serializable
internal data class CreatedApiKeyDto(val apiKey: ApiKeyDto, val key: String)

@Serializable
internal data class ApiKeyPageDto(val items: List<ApiKeyDto> = emptyList(), val nextCursor: String? = null)

@Serializable
internal data class CreateApiKeyRequestDto(
    val name: String? = null,
    val scopes: List<String>,
    val expiresInSeconds: Long? = null,
)

@Serializable
internal data class ProblemDto(
    val type: String? = null,
    val title: String? = null,
    val status: Int? = null,
    val detail: String? = null,
    val correlationId: String? = null,
    val retryAfterSeconds: Long? = null,
    val category: String? = null,
    val quota: String? = null,
    val limit: Long? = null,
    val current: Long? = null,
    val apiKeyId: String? = null,
    val publicId: String? = null,
)

/**
 * Identifies this SDK in the server's access logs.
 *
 * Until now no shipped SDK sent one, so requests were attributed to whatever
 * the runtime defaulted to. Staging Ops found the consequence: reaching the
 * deployed API depended on an unexamined interaction between a runtime default
 * and a Cloudflare heuristic — harmless in practice, but nothing would have
 * predicted or quickly diagnosed it if it had not been.
 *
 * Kept in step with the Gradle version by `SdkVersionTest`.
 */
internal const val SDK_VERSION = "0.1.0-SNAPSHOT"
internal const val USER_AGENT = "testinbox-sdk-jvm/$SDK_VERSION"

internal class Transport(
    private val baseUrl: String,
    private val apiKey: String,
    private val httpClient: HttpClient = HttpClient.newHttpClient(),
) {
    // Forward compatibility: unknown fields and enum values must never break parsing.
    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    suspend fun createInbox(request: CreateInboxRequestDto, idempotencyKey: String? = null): InboxDto =
        json.decodeFromString(
            InboxDto.serializer(),
            execute(
                "POST",
                "/v1/inboxes",
                json.encodeToString(CreateInboxRequestDto.serializer(), request),
                idempotencyKey,
            ).let { String(it) },
        )

    suspend fun getInbox(id: String): InboxDto =
        json.decodeFromString(InboxDto.serializer(), String(execute("GET", "/v1/inboxes/$id")))

    suspend fun deleteInbox(id: String) {
        execute("DELETE", "/v1/inboxes/$id")
    }

    suspend fun listMessages(inboxId: String): MessagePageDto =
        json.decodeFromString(
            MessagePageDto.serializer(),
            String(execute("GET", "/v1/inboxes/$inboxId/messages")),
        )

    suspend fun wait(inboxId: String, request: WaitRequestDto): WaitResultDto =
        json.decodeFromString(
            WaitResultDto.serializer(),
            String(
                execute(
                    "POST",
                    "/v1/inboxes/$inboxId/messages/wait",
                    json.encodeToString(WaitRequestDto.serializer(), request),
                ),
            ),
        )

    suspend fun getMessage(id: String): MessageDto =
        json.decodeFromString(MessageDto.serializer(), String(execute("GET", "/v1/messages/$id")))

    suspend fun rawMime(messageId: String): ByteArray = execute("GET", "/v1/messages/$messageId/raw")

    suspend fun createApiKey(
        request: CreateApiKeyRequestDto,
        idempotencyKey: String? = null,
    ): CreatedApiKeyDto =
        json.decodeFromString(
            CreatedApiKeyDto.serializer(),
            String(
                execute(
                    "POST",
                    "/v1/api-keys",
                    json.encodeToString(CreateApiKeyRequestDto.serializer(), request),
                    idempotencyKey,
                ),
            ),
        )

    suspend fun listApiKeys(cursor: String?, limit: Int?): ApiKeyPageDto {
        val query =
            listOfNotNull(
                cursor?.let { "cursor=" + java.net.URLEncoder.encode(it, Charsets.UTF_8) },
                limit?.let { "limit=$it" },
            ).joinToString("&")
        return json.decodeFromString(
            ApiKeyPageDto.serializer(),
            String(execute("GET", "/v1/api-keys" + if (query.isEmpty()) "" else "?$query")),
        )
    }

    suspend fun getApiKey(id: String): ApiKeyDto =
        json.decodeFromString(ApiKeyDto.serializer(), String(execute("GET", "/v1/api-keys/$id")))

    suspend fun revokeApiKey(id: String) {
        execute("DELETE", "/v1/api-keys/$id")
    }

    private suspend fun execute(
        method: String,
        path: String,
        body: String? = null,
        idempotencyKey: String? = null,
    ): ByteArray {
        val builder =
            HttpRequest.newBuilder(URI.create(baseUrl.trimEnd('/') + path))
                .header("Authorization", "Bearer $apiKey")
                .header("Accept", "application/json, message/rfc822, application/octet-stream")
                .header("User-Agent", USER_AGENT)
        // Sent verbatim; never generated or rewritten here (ADR-033).
        idempotencyKey?.let { builder.header("Idempotency-Key", it) }
        if (body != null) {
            builder.header("Content-Type", "application/json")
            builder.method(method, HttpRequest.BodyPublishers.ofString(body))
        } else {
            builder.method(method, HttpRequest.BodyPublishers.noBody())
        }
        val response =
            httpClient.sendAsync(builder.build(), HttpResponse.BodyHandlers.ofByteArray()).await()
        if (response.statusCode() in 200..299) return response.body()
        throw mapError(response.statusCode(), response.body(), response.headers())
    }

    private fun mapError(
        status: Int,
        body: ByteArray,
        headers: java.net.http.HttpHeaders,
    ): RuntimeException {
        val problem =
            runCatching { json.decodeFromString(ProblemDto.serializer(), String(body)) }
                .getOrElse { ProblemDto() }
        val detail = problem.detail ?: problem.title ?: "HTTP $status"
        val retryAfter =
            (problem.retryAfterSeconds ?: headers.firstValue("Retry-After").orElse(null)?.toLongOrNull())
                ?.let(java.time.Duration::ofSeconds)
        return when {
            status == 401 -> TestInboxAuthException(detail, problem.correlationId, problem.type, status)
            // 403 carries two meanings — `missing-scope` and
            // `scope-escalation` — so the problem type travels with it.
            status == 403 -> TestInboxForbiddenException(detail, problem.correlationId, problem.type, status)
            status == 404 -> TestInboxNotFoundException(detail, problem.correlationId, problem.type, status)
            // Two distinct 409s share this status (ADR-021 vs ADR-027), so the
            // problem type — not the status code — decides which error this is.
            status == 409 && problem.type?.endsWith("/quota-exceeded") == true ->
                TestInboxQuotaExceededException(
                    detail,
                    problem.correlationId,
                    quota = problem.quota,
                    limit = problem.limit,
                    current = problem.current,
                )
            // Several distinct 409s share this status (ADR-021, ADR-027,
            // ADR-033) and their correct client actions differ — free
            // capacity, wait out a cooldown, retry with the same key, never
            // reuse the key, or revoke and re-mint. Discriminating on the
            // problem type is what docs/api/principles.md #3 asks of a client.
            status == 409 && problem.type?.endsWith("/idempotency-request-in-progress") == true ->
                TestInboxIdempotencyInProgressException(detail, problem.correlationId, retryAfter)
            status == 409 && problem.type?.endsWith("/idempotency-secret-not-replayable") == true ->
                TestInboxCredentialAlreadyCreatedException(
                    detail,
                    problem.correlationId,
                    problem.apiKeyId,
                    problem.publicId,
                )
            status == 409 &&
                (
                    problem.type?.endsWith("/idempotency-key-reused") == true ||
                        problem.type?.endsWith("/idempotency-replay-unavailable") == true
                ) ->
                // The two terminal types share one exception because they ask
                // the same thing of a caller, but `problemType` is carried so
                // "my key scheme is wrong" stays distinguishable from "an
                // artifact rollback ate my record".
                TestInboxIdempotencyConflictException(detail, problem.correlationId, problem.type)
            status == 409 -> TestInboxConflictException(detail, problem.correlationId, problem.retryAfterSeconds)
            status == 410 -> TestInboxInboxGoneException(detail, problem.correlationId, problem.type, status)
            status == 429 ->
                TestInboxRateLimitException(
                    detail,
                    problem.correlationId,
                    retryAfter = retryAfter,
                    category = problem.category,
                    limit = problem.limit ?: headers.firstValue("RateLimit-Limit").orElse(null)?.toLongOrNull(),
                    remaining = headers.firstValue("RateLimit-Remaining").orElse(null)?.toLongOrNull(),
                )
            else -> TestInboxApiException(status, problem.type, detail, problem.correlationId)
        }
    }
}

File: sdk/typescript/src/internal/transport.ts
/**
 * INTERNAL transport layer — thin, typed wrappers over the v1 REST contract
 * (`backend/api/contract/openapi.yaml`).
 *
 * Per ADR-014 this module is an implementation detail: it is NOT exported from
 * the package entry point and nothing here is part of the public API. It may
 * be swapped (e.g. for a generated client) without a public API break.
 *
 * Wire DTOs deliberately carry `[key: string]: unknown` index signatures and
 * optional enum-ish `string` fields: unknown fields and unknown enum values
 * from a newer server must never cause a failure (docs/sdk/principles.md #5).
 */

import {
  TestInboxApiError,
  TestInboxQuotaExceededError,
  TestInboxRateLimitError,
  TestInboxAuthError,
  TestInboxConflictError,
  TestInboxError,
  TestInboxForbiddenError,
  TestInboxInboxGoneError,
  TestInboxNotFoundError,
  type ProblemDetails,
} from "../errors";

// ---------------------------------------------------------------------------
// Wire DTOs (request/response shapes of the REST contract)
// ---------------------------------------------------------------------------

export interface CreateInboxRequestDto {
  addressMode?: string;
  ttlSeconds?: number;
  aliasHint?: string;
  localPart?: string;
}

export interface InboxDto {
  id: string;
  address: string;
  addressMode?: string;
  state?: string;
  createdAt?: string;
  expiresAt?: string;
  [key: string]: unknown;
}

export interface EmailHeaderDto {
  name: string;
  value: string;
  [key: string]: unknown;
}

export interface EmailLinkDto {
  href: string;
  text?: string;
  [key: string]: unknown;
}

export interface AttachmentMetaDto {
  id: string;
  fileName?: string;
  contentType?: string;
  sizeBytes?: number;
  [key: string]: unknown;
}

export interface MessageDto {
  id: string;
  inboxId?: string;
  receivedAt?: string;
  parseStatus?: string;
  from?: string;
  subject?: string;
  textBody?: string;
  htmlBody?: string;
  links?: EmailLinkDto[];
  headers?: EmailHeaderDto[];
  attachments?: AttachmentMetaDto[];
  [key: string]: unknown;
}

export interface MessagePageDto {
  items?: MessageDto[];
  nextCursor?: string | null;
  [key: string]: unknown;
}

export interface HeaderMatcherDto {
  name: string;
  value?: string;
}

export interface MessageMatcherDto {
  from?: string;
  subjectContains?: string;
  subjectEquals?: string;
  headers?: HeaderMatcherDto[];
}

export interface WaitRequestDto {
  matcher?: MessageMatcherDto;
  timeoutSeconds: number;
}

/**
 * Wait outcome. `status` is an open string: the contract requires clients to
 * tolerate unknown future status values (treated by the caller as "keep
 * chaining", ADR-020).
 */
export interface WaitResultDto {
  status?: string;
  message?: MessageDto;
  elapsedMs?: number;
  arrivedButUnmatchedCount?: number;
  parseFailedCount?: number;
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Error mapping (RFC 7807 -> typed hierarchy)
// ---------------------------------------------------------------------------

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asProblemDetails(status: number, body: unknown): ProblemDetails {
  if (!isRecord(body)) return { status };
  return {
    status,
    type: typeof body.type === "string" ? body.type : undefined,
    title: typeof body.title === "string" ? body.title : undefined,
    detail: typeof body.detail === "string" ? body.detail : undefined,
    instance: typeof body.instance === "string" ? body.instance : undefined,
    correlationId: typeof body.correlationId === "string" ? body.correlationId : undefined,
    retryAfterSeconds:
      typeof body.retryAfterSeconds === "number" ? body.retryAfterSeconds : undefined,
    category: typeof body.category === "string" ? body.category : undefined,
    quota: typeof body.quota === "string" ? body.quota : undefined,
    limit: typeof body.limit === "number" ? body.limit : undefined,
    current: typeof body.current === "number" ? body.current : undefined,
  };
}

function errorForStatus(status: number, problem: ProblemDetails): TestInboxError {
  const summary = problem.title
    ? `${problem.title}${problem.detail ? `: ${problem.detail}` : ""}`
    : `request failed`;
  const message = `TestInbox API error (HTTP ${status}): ${summary}`;
  switch (status) {
    case 401:
      return new TestInboxAuthError(message, problem);
    case 403:
      return new TestInboxForbiddenError(message, problem);
    case 404:
      return new TestInboxNotFoundError(message, problem);
    case 409:
      // Two distinct 409 meanings share this status (ADR-021 vs ADR-027), so
      // the problem type decides which error this is — never the status code.
      return problem.type?.endsWith("/quota-exceeded")
        ? new TestInboxQuotaExceededError(message, problem)
        : new TestInboxConflictError(message, problem);
    case 410:
      return new TestInboxInboxGoneError(message, problem);
    case 429:
      return new TestInboxRateLimitError(message, problem);
    default:
      return new TestInboxApiError(message, problem);
  }
}

// ---------------------------------------------------------------------------
// Transport
// ---------------------------------------------------------------------------

export interface TransportOptions {
  apiKey: string;
  baseUrl: string;
}

interface RequestOptions {
  body?: unknown;
  accept?: string;
  signal?: AbortSignal;
}

/**
 * Minimal HTTP client over the global `fetch` (Node >= 20; zero runtime
 * dependencies). Holds the API key in a private field so it can never be
 * enumerated, logged, or JSON-serialized.
 */
export class Transport {
  readonly #apiKey: string;
  readonly #baseUrl: string;

  constructor(options: TransportOptions) {
    this.#apiKey = options.apiKey;
    this.#baseUrl = options.baseUrl.replace(/\/+$/, "");
  }

  async createInbox(request: CreateInboxRequestDto): Promise<InboxDto> {
    const res = await this.#request("POST", "/v1/inboxes", { body: request });
    return (await res.json()) as InboxDto;
  }

  async getInbox(inboxId: string): Promise<InboxDto> {
    const res = await this.#request("GET", `/v1/inboxes/${encodeURIComponent(inboxId)}`, {});
    return (await res.json()) as InboxDto;
  }

  async deleteInbox(inboxId: string): Promise<void> {
    await this.#request("DELETE", `/v1/inboxes/${encodeURIComponent(inboxId)}`, {});
  }

  async listMessages(inboxId: string, cursor?: string): Promise<MessagePageDto> {
    const query = cursor ? `?cursor=${encodeURIComponent(cursor)}` : "";
    const res = await this.#request(
      "GET",
      `/v1/inboxes/${encodeURIComponent(inboxId)}/messages${query}`,
      {},
    );
    return (await res.json()) as MessagePageDto;
  }

  async wait(inboxId: string, request: WaitRequestDto): Promise<WaitResultDto> {
    // Safety net: abort only well past the server's own window, so a hung
    // connection cannot stall the caller indefinitely. Normal wait-window
    // expiry is a 200 {status: TIMEOUT} response, never an abort.
    const signal = AbortSignal.timeout((request.timeoutSeconds + 15) * 1000);
    const res = await this.#request(
      "POST",
      `/v1/inboxes/${encodeURIComponent(inboxId)}/messages/wait`,
      { body: request, signal },
    );
    return (await res.json()) as WaitResultDto;
  }

  async getRawMime(messageId: string): Promise<Uint8Array> {
    const res = await this.#request("GET", `/v1/messages/${encodeURIComponent(messageId)}/raw`, {
      accept: "message/rfc822",
    });
    return new Uint8Array(await res.arrayBuffer());
  }

  async #request(method: string, path: string, options: RequestOptions): Promise<Response> {
    const headers: Record<string, string> = {
      authorization: `Bearer ${this.#apiKey}`,
      accept: options.accept ?? "application/json, application/problem+json",
    };
    let body: string | undefined;
    if (options.body !== undefined) {
      headers["content-type"] = "application/json";
      body = JSON.stringify(options.body);
    }

    let response: Response;
    try {
      response = await fetch(this.#baseUrl + path, {
        method,
        headers,
        body,
        signal: options.signal,
      });
    } catch (cause) {
      const reason = cause instanceof Error ? cause.message : String(cause);
      throw new TestInboxError(`TestInbox API request failed: ${reason}`, undefined, { cause });
    }

    if (!response.ok) {
      throw errorForStatus(response.status, asProblemDetails(response.status, await safeJson(response)));
    }
    return response;
  }
}

async function safeJson(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch {
    return undefined;
  }
}
